### PR TITLE
DMP-3645 - Prevent expired case URLs from being accessed

### PR DESCRIPTION
**Pull request by Ben-Edwards-cgi**
### Change description ###
When DMP-2909 is implemented, only basic information about a case will be displayed to the user. DARTS should prevent users from manually trying to change URLs to access the tabs that have been hidden.

When a user tries to access hidden parts of the case with the *is_data_anonymised* flag set to true, return a 404 error with the following detail:
 * code = CASE_XXX (replacing XXX with the next number in the sequence)
 * title = Case has expired

This needs to be applied to all DARTS portal endpoints that can be called from the frontend for a case e.g. Hearings, Transcripts, Annotations, Events, Audios etc.

The admin endpoints should remain accessible. 

The frontend part of this ticket can be found in DMP-2989. 

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/darts/cases/service/impl/CaseServiceImpl.java [hunk 1]
```diff
@@ -143,15 +145,15 @@ public PostCaseResponse addCaseOrUpdate(AddCaseRequest addCaseRequest) {
         CourtCaseEntity courtCase = retrieveCoreObjectService.retrieveOrCreateCase(
             addCaseRequest.getCourthouse(),
             addCaseRequest.getCaseNumber()
-        );
+        ).validateIsExpired();
```
Comment (davet1985) [inline src/main/java/uk/gov/hmcts/darts/cases/service/impl/CaseServiceImpl.java:148]: I'm not sure about adding this here because this is used by the "addCase" endpoint and we don't want to expose this case expired error to the external services which call that endpoint. 

I think we should ensure this validation is only enforced for endpoints which are used by the portal.

File: src/main/java/uk/gov/hmcts/darts/cases/service/impl/CaseServiceImpl.java [hunk 2]
```diff
@@ -143,15 +145,15 @@ public PostCaseResponse addCaseOrUpdate(AddCaseRequest addCaseRequest) {
...
         return updateCase(addCaseRequest, courtCase);
     }
 
     private PostCaseResponse updateCase(AddCaseRequest addCaseRequest, CourtCaseEntity existingCase) {
         CourtCaseEntity updatedCaseEntity = casesMapper.addDefendantProsecutorDefenderJudgeType(
             existingCase,
             addCaseRequest
-        );
+        ).validateIsExpired();
```
Comment (davet1985) [inline src/main/java/uk/gov/hmcts/darts/cases/service/impl/CaseServiceImpl.java:156]: Same as above.

File: src/main/java/uk/gov/hmcts/darts/common/service/impl/CaseCommonServiceImpl.java [hunk 3]
```diff
@@ -25,16 +25,20 @@ public class CaseCommonServiceImpl implements CaseCommonService {
     public CourtCaseEntity retrieveOrCreateCase(String courthouseName, String caseNumber, UserAccountEntity userAccount) {
         String courthouseNameUC = StringUtils.toRootUpperCase(courthouseName);
         Optional<CourtCaseEntity> foundCase = caseRepository.findByCaseNumberAndCourthouse_CourthouseName(caseNumber, courthouseNameUC);
-        return foundCase.map(entity -> setCourtCaseLastDateModifiedBy(entity, userAccount))
-            .orElseGet(() -> createCase(courthouseName, caseNumber, userAccount));
+        return foundCase
+            .map(entity -> setCourtCaseLastDateModifiedBy(entity, userAccount))
+            .orElseGet(() -> createCase(courthouseName, caseNumber, userAccount))
+            .validateIsExpired();
```
Comment (davet1985) [inline src/main/java/uk/gov/hmcts/darts/common/service/impl/CaseCommonServiceImpl.java:31]: As per other comment, we should keep this validation out of common code that fetches the case.

File: src/main/java/uk/gov/hmcts/darts/common/service/impl/CaseCommonServiceImpl.java [hunk 4]
```diff
@@ -25,16 +25,20 @@ public class CaseCommonServiceImpl implements CaseCommonService {
...
     @Override
     @Transactional
     public CourtCaseEntity retrieveOrCreateCase(CourthouseEntity courthouse, String caseNumber, UserAccountEntity userAccount) {
         Optional<CourtCaseEntity> foundCase = caseRepository.findByCaseNumberAndCourthouse(caseNumber, courthouse);
-        return foundCase.map(entity -> setCourtCaseLastDateModifiedBy(entity, userAccount))
-            .orElseGet(() -> createCase(courthouse, caseNumber, userAccount));
+        return foundCase
+            .map(entity -> setCourtCaseLastDateModifiedBy(entity, userAccount))
+            .orElseGet(() -> createCase(courthouse, caseNumber, userAccount))
+            .validateIsExpired();
```
Comment (davet1985) [inline src/main/java/uk/gov/hmcts/darts/common/service/impl/CaseCommonServiceImpl.java:41]: Same as above.

File: src/main/java/uk/gov/hmcts/darts/task/service/impl/AutomatedTaskServiceImpl.java [hunk 5]
```diff
@@ -56,8 +57,15 @@ public class AutomatedTaskServiceImpl implements AutomatedTaskService {
...
 
+    @Value("${darts.automated.allow-automated-tasks:true}")
+    private final boolean allowAutomatedTasks;
+
     @Override
     public void configureAndLoadAutomatedTasks(ScheduledTaskRegistrar taskRegistrar) {
+        if (!allowAutomatedTasks) {
+            log.info("Automated tasks are disabled");
+            return;
+        }
```
Comment (davet1985) [inline src/main/java/uk/gov/hmcts/darts/task/service/impl/AutomatedTaskServiceImpl.java:68]: Is this change intentional? It doesn't seem relevant for the ticket.

File: src/main/resources/application.yaml [hunk 6]
```diff
@@ -83,7 +83,7 @@ spring:
       maximum-pool-size: ${DARTS_API_DB_POOL_SIZE:50}
   jpa:
     database: postgresql
-    show-sql: false
+    show-sql: true
```
Comment (davet1985) [inline src/main/resources/application.yaml:86]: Not sure we want to turn this on everywhere? Maybe set in a local profile?

**Review by davet1985 — CHANGES_REQUESTED** [review]
I'm concerned about adding this validation in some very common methods called by code that we don't want to expose the error too.

File: src/main/resources/application.yaml (hunk 6 above)
Comment (Ben-Edwards-cgi) [reply src/main/resources/application.yaml:86]: Agreed this was added temporarily and should have been removed - Updated

File: src/main/java/uk/gov/hmcts/darts/task/service/impl/AutomatedTaskServiceImpl.java (hunk 5 above)
Comment (Ben-Edwards-cgi) [reply src/main/java/uk/gov/hmcts/darts/task/service/impl/AutomatedTaskServiceImpl.java:68]: I spoke to Jack on this one and he is happy as it makes local testing easier if these are disabled (It will be enabled on all deployed envs by default)

File: src/main/java/uk/gov/hmcts/darts/cases/service/impl/CaseServiceImpl.java (hunk 1 above)
Comment (Ben-Edwards-cgi) [reply src/main/java/uk/gov/hmcts/darts/cases/service/impl/CaseServiceImpl.java:148]: Updated

File: src/main/java/uk/gov/hmcts/darts/cases/service/impl/CaseServiceImpl.java (hunk 2 above)
Comment (Ben-Edwards-cgi) [reply src/main/java/uk/gov/hmcts/darts/cases/service/impl/CaseServiceImpl.java:156]: Updated

File: src/main/java/uk/gov/hmcts/darts/common/service/impl/CaseCommonServiceImpl.java (hunk 3 above)
Comment (Ben-Edwards-cgi) [reply src/main/java/uk/gov/hmcts/darts/common/service/impl/CaseCommonServiceImpl.java:31]: Updated

File: src/main/java/uk/gov/hmcts/darts/common/service/impl/CaseCommonServiceImpl.java (hunk 4 above)
Comment (Ben-Edwards-cgi) [reply src/main/java/uk/gov/hmcts/darts/common/service/impl/CaseCommonServiceImpl.java:41]: Updated

**Review by davet1985 — APPROVED** [review]
